File: frontend/src/app/modules/teacher/services/teacher.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { API_CONFIG } from '../../../core/config/api.config';

const API_URL = API_CONFIG.TEACHER + '/';
const ADMIN_API_URL = API_CONFIG.ADMIN + '/'; // For updating profile

@Injectable({
  providedIn: 'root'
})
export class TeacherService {

  constructor(private http: HttpClient) { }

  getTeacherDashboard(): Observable<any> {
    return this.http.get(API_URL + 'dashboard');
  }

  getTeacherProfile(): Observable<any> {
    return this.http.get(API_URL + 'profile');
  }

  // A teacher can edit their own profile (we use the admin endpoint for this)
  updateProfile(teacherId: number, teacherData: any): Observable<any> {
    return this.http.put(ADMIN_API_URL + `teachers/${teacherId}`, teacherData);
  }

  addOrUpdateAttendance(attendanceData: any): Observable<any> {
    return this.http.post(API_URL + 'attendance', attendanceData);
  }

  addOrUpdateMarks(marksData: any): Observable<any> {
    return this.http.post(API_URL + 'marks', marksData);
  }

  generateReport(studentId: number): Observable<Blob> {
    // We expect a PDF file (Blob) as the response
    return this.http.get(API_URL + `reports/student/${studentId}`, { responseType: 'blob' });
  }

  getStudentAttendance(studentId: number): Observable<any> {
    return this.http.get(API_URL + `attendance/student/${studentId}`);
  }

  getStudentMarks(studentId: number): Observable<any> {
    return this.http.get(API_URL + `marks/student/${studentId}`);
  }

  getAllStudentsAttendance(): Observable<any> {
    return this.http.get(API_URL + 'attendance/all');
  }

  getAllStudentsMarks(): Observable<any> {
    return this.http.get(API_URL + 'marks/all');
  }

  getAllSubjects(): Observable<any> {
    return this.http.get(API_URL + 'subjects');
  }

  deleteAttendance(attendanceId: number): Observable<any> {
    return this.http.delete(API_URL + `attendance/${attendanceId}`, { responseType: 'text' });
  }

  deleteMarks(marksId: number): Observable<any> {
    return this.http.delete(API_URL + `marks/${marksId}`, { responseType: 'text' });
  }
}

File: backend/src/main/java/com/school/management/system/service/TeacherService.java
package com.school.management.system.service;

import com.school.management.system.entity.Attendance;
import com.school.management.system.entity.Marks;
import com.school.management.system.entity.Subject;
import com.school.management.system.entity.Teacher;
import com.school.management.system.payload.request.AttendanceRequestDTO;
import com.school.management.system.payload.request.MarksRequestDTO;
import com.school.management.system.payload.response.TeacherDashboardDTO;

import java.util.List;

public interface TeacherService {
 
    TeacherDashboardDTO getTeacherDashboard();
    Teacher getTeacherProfile();
    Attendance markStudentAttendance(AttendanceRequestDTO attendanceDTO);
    Marks addOrUpdateStudentMarks(MarksRequestDTO marksDTO);
    byte[] generatePerformanceReport(Long studentId);
    List<Attendance> getAllStudentsAttendance();
    List<Marks> getAllStudentsMarks();
    List<Attendance> getStudentAttendance(Long studentId);
    List<Marks> getStudentMarks(Long studentId);
    List<Subject> getAllSubjects();
    void deleteAttendance(Long attendanceId);
    void deleteMarks(Long marksId);
}